src/toil/test/server/serverTest.py

```diff
@@ -0,0 +1,136 @@
+# Copyright (C) 2015-2021 Regents of the University of California
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import json
+import logging
+import os
+import textwrap
+import unittest
+from io import BytesIO
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from flask import Flask
+
+from toil.test import ToilTest, needs_server
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+
+@needs_server
+class ToilServerUtilsTest(ToilTest):
+    """
+    Tests for the utility functions used by the Toil server.
+    """
+
+
+@needs_server
+class ToilWESServerTest(ToilTest):
+    """
+    Tests for Toil's Workflow Execution Service API support using Flask's
+    builtin test client.
+    """
+
+    def setUp(self) -> None:
+        super(ToilWESServerTest, self).setUp()
+        self.temp_dir = self._createTempDir()
+
+        from toil.server.app import create_app, parser_with_server_options
+        parser = parser_with_server_options()
+        args = parser.parse_args(["--work_dir", os.path.join(self.temp_dir, "workflows")])
+
+        self.app: "Flask" = create_app(args).app
```

Comment:
Is this smart enough to make sure that only localhost can access this test server?

Reply:
This is using Flask's testing API where the requests are routed internally, so no server is actually started.